internal/db/db.go

```diff
@@ -49,57 +61,116 @@ func (d *Database) Transaction(f func(*Database) error) error {
-	if err := db.Exec(string(schema)).Error; err != nil {
-		return errors.E(op, dbError(err))
+
+	sqlDB, err := db.DB()
+	if err != nil {
+		return errors.E(op, fmt.Errorf("unable to obtain raw DB: %w", err))
+	}
+	conn, err := sqlDB.Conn(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to obtain DB conn: %w", err)
 	}
 
-	for {
-		v := dbmodel.Version{Component: dbmodel.Component, Major: 1, Minor: 0}
-		if err := db.FirstOrCreate(&v).Error; err != nil {
-			return errors.E(op, dbError(err))
-		}
-		if dbmodel.Major == v.Major && dbmodel.Minor <= v.Minor {
-			// The database is already at, or past, our current version.
-			// Nothing to do.
-			atomic.StoreUint32(&d.migrated, 1)
-			return nil
-		}
-		if v.Major != dbmodel.Major && !force {
-			return errors.E(op, errors.CodeServerConfiguration, fmt.Sprintf("database has incompatible version %d.%d", v.Major, v.Minor))
+	sqlDir, err := iofs.New(dbmodel.SQL, path.Join("sql", db.Name()))
+	if err != nil {
+		return errors.E(op, fmt.Errorf("unable to create new sql filesys: %w", err))
+	}
+
+	driver, err := postgres.WithConnection(ctx, conn, &postgres.Config{MigrationsTable: MigrationTableName})
+	if err != nil {
+		return errors.E(op, fmt.Errorf("unable to create new driver instance: %w", err))
+	}
+
+	// DB name is left blank because it is contained in the driver/DB connection.
+	m, err := migrate.NewWithInstance("iofs", sqlDir, "", driver)
+	if err != nil {
+		return errors.E(op, fmt.Errorf("unable to create new migrator: %w", err))
+	}
+	defer m.Close()
+
+	// Setup custom logger for consistent output.
+	logger := migrationLogger{logger: zapctx.Logger(ctx), verbose: false}
+	m.Log = logger
+
+	if err := d.handleDeprecatedMigrations(ctx, m); err != nil {
+		return errors.E(op, fmt.Errorf("failed to handle deprecated migrations: %w", err))
+	}
+
+	v, dirty, err := m.Version()
+	if err != nil {
+		if !stderr.Is(err, migrate.ErrNilVersion) {
+			return errors.E(op, fmt.Errorf("failed to get db version: %w", err))
 		}
-		// The major versions are unchanged, the database can be migrated.
-		v.Minor += 1
-		schema, err := dbmodel.SQL.ReadFile(path.Join("sql", db.Name(), fmt.Sprintf("%d_%d.sql", v.Major, v.Minor)))
-		if err != nil {
-			return errors.E(op, err)
+	}
+
+	if dirty {
+		// nolint:gosec
+		workingVersion := int(v) - 1
+		zapctx.Info(ctx, "dirty database, reverting version", zap.Int("version", workingVersion))
+		if err := m.Force(workingVersion); err != nil {
+			return errors.E(op, fmt.Errorf("failed to fix dirty db version: %w", err))
 		}
+	}
 
-		err = db.Transaction(func(tx *gorm.DB) error {
-			if err := tx.Exec(string(schema)).Error; err != nil {
-				return err
-			}
-			return nil
-		})
-		if err != nil {
-			return errors.E(op, dbError(err))
+	if err := m.Up(); err != nil {
+		if !stderr.Is(err, migrate.ErrNoChange) {
+			return errors.E(op, fmt.Errorf("failed to migrate db: %w", err))
 		}
 	}
+
+	atomic.StoreUint32(&d.migrated, 1)
```

Comment:
why don't we also check this flag at the start of the `Migrate` method?

Reply:
We could, but we didn't before so I haven't changed that logic.

Reply:
i guess it's uncommon we call the `Migrate` twice in our tests, because we usually create a new db for each test.